### PR TITLE
[YUNIKORN-1647] Dynamic namespace support in admission controller

### PR DESCRIPTION
**Pull request by wilfred-s**
### What is this PR for?
Support annotations on a namespace to define processing behaviour in the admission controller. The annotations have preference over the regexp. If no annotations are found fall back on the regexp. If an annotation is found it defines the behaviour and overrides what is defined in the regexp.

Annotations to be supported on a namespace:
* yunikorn.apache.org/namespace.generateAppId
* yunikorn.apache.org/namespace.enableYuniKorn

Supported values for the annotation: true and false.

### What type of PR is it?
* [X] - Feature

### What is the Jira issue?
* https://issues.apache.org/jira/browse/YUNIKORN-1647

### How should this be tested?
Unit tests are added.

### CODE REVIEW

**Review by craigcondit — APPROVED** [review]
Not entirely thrilled with the "int as tri-state bool" pattern... It just feels ugly. An enum with constants might be cleaner, and easier to follow in the logic. True/False/Maybe. We get type safety that way too.

File: pkg/admission/admission_controller.go [hunk 1]
```diff
@@ -524,11 +526,28 @@ func (c *AdmissionController) namespaceMatchesNoLabelList(namespace string) bool
...
 
+// shouldProcessNamespace returns true if the pod in the namespace must be redirected to the
+// YuniKorn scheduler by setting the schedulerName
+// First check is the namespace annotation (tri-state)
+// - if present (0, 1) return the value as boolean
+// - if not present (-1) fallback to matching names based on regexp
 func (c *AdmissionController) shouldProcessNamespace(namespace string) bool {
+	process := c.nsCache.enableYuniKorn(namespace)
+	if process != -1 {
+		return process == 1
```
Comment (pbacsko) [inline pkg/admission/admission_controller.go:537]: 1 / -1 are special values, could you name them? Difficult to reason about them without names.

File: pkg/admission/admission_controller.go [hunk 2]
```diff
@@ -524,11 +526,28 @@ func (c *AdmissionController) namespaceMatchesNoLabelList(namespace string) bool
...
 
+// shouldLabelNamespace returns true if the pod in the namespace must be labeled
+// First check is the namespace annotation (tri-state)
+// - if present (0, 1) return the value as boolean
+// - if not present (-1) fallback to matching names based on regexp
 func (c *AdmissionController) shouldLabelNamespace(namespace string) bool {
+	label := c.nsCache.generateAppID(namespace)
+	if label != -1 {
+		return label == 1
+	}
```
Comment (pbacsko) [inline pkg/admission/admission_controller.go:550]: Same thing here

**Comment by pbacsko** [comment]
> Not entirely thrilled with the "int as tri-state bool" pattern... It just feels ugly. An enum with constants might be cleaner, and easier to follow in the logic. True/False/Maybe. We get type safety that way too.

+1 on this, naming the values increases readability.

File: pkg/admission/admission_controller.go (hunk 1 above)
Comment (wilfred-s) [reply pkg/admission/admission_controller.go:537]: Follow up jira to move to an iota: [YUNIKORN-1652](https://issues.apache.org/jira/browse/YUNIKORN-1653)

**Pull request closed without merging**